### PR TITLE
first version ECS task definition

### LINKED ISSUE #18: Wordpress Application
**Task definition:**
- custom wordpress image from ECR
- using efs as volume and mounting /var/www/html inside wordpress container for persistent storage
  - no need for IAM authentication
- wordpress database variables coming from ssm parameter store
- memory 128
- any IAM execution role if needed with only required permissions

**ecs service**

- type EC2
- use custom task definition for wordpress
- service type replica
- deployment type rolling update or blue/green (first is easier, second is better)
- at least 2 task definitions running at any point in time

PS: need to initialise wordpress database when first run a container

**DOD**

- for testing purposes:
  - run a test ecs cluster with instances running on public subnet and required security groups open
  - run service with only one task definition running exposed to port 80
  - access public ip address from ec2 instance on port 80 to reach wordpress running
  - finish wordpress installation steps to initialise database
- terraform code:
  - service needs 2 task definitions running
  - don’t expose task definition to port 80 on host

### PR DESCRIPTION
**Pull request by CarineOliveira**

### CODE REVIEW

**Review by hsy3418 — CHANGES_REQUESTED** [review]
I think we also need db_host,db_name,db_password,db_user to the task definitions, and also need to create an ecs service for deploying the task definitions.

File: wordpress-application/task-definition.tf [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      "name": "wordpress-container",
+      "image": "160372566358.dkr.ecr.ap-southeast-2.amazonaws.com/group3-wordpress:latest",
+      "cpu": 10,
+      "memory": 512,
+      "essential": true,
+
+      "secrets": [
+        { 
+          "name": "WORDPRESS_DB_HOST",
+          "valueFrom": "arn:aws:ssm:ap-southeast-2:160372566358:parameter/WORDPRESS_DB_HOST"
```
Comment (caiocezart) [inline wordpress-application/task-definition.tf]: make the account number a variable. you can get that from a data resource
same for region

`data "aws_caller_identity" "current" {}`
`data "aws_region" "current" {}`

`arn:aws:ssm:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:parameter/WORDPRESS_DB_HOST`

Comment (caiocezart) [reply wordpress-application/task-definition.tf]: do the same for other secrets

File: wordpress-application/task-definition.tf [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        {
+          "containerPort": 80,
+          "hostPort": 80
+        }
+      ]
+    }
+  ]
+  TASK_DEFINITION
+
+  volume {
```
Comment (caiocezart) [inline wordpress-application/task-definition.tf:61]: to use this we need the access point which wasn't available in terraform, remember?
The other way of doing is:
```
{
    "containerDefinitions": [
        {
            "mountPoints": [
                {
                    "sourceVolume": "wordpress",
                    "containerPath": "/var/www/html"
                }
            ]
        }
    ],
    "volumes": [
        {
            "name": "wordpress",
            "host": {
                "sourcePath": "/mnt/efs/wordpress"
            }
        }
    ]
}
```

replace the values as per efs configuration, use variable as needed

File: wordpress-application/terraform.tfvars [hunk 3]
```diff
@@ -0,0 +1,16 @@
+region = "ap-southeast-2"
```
Comment (caiocezart) [inline wordpress-application/terraform.tfvars]: no need for var region, see example above with data resources

File: wordpress-application/terraform.tfvars [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+aws_account_id = "160372566358"
```
Comment (caiocezart) [inline wordpress-application/terraform.tfvars]: same here

File: wordpress-application/variables.tf [hunk 5]
```diff
@@ -0,0 +1,29 @@
+variable "region" {
```
Comment (caiocezart) [inline wordpress-application/variables.tf]: same

File: wordpress-application/variables.tf [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  type = "string"
+}
+
+variable "aws_account_id" {
```
Comment (caiocezart) [inline wordpress-application/variables.tf]: same

**Review by caiocezart — CHANGES_REQUESTED** [review]
good work, a few notes

File: wordpress-application/task-definition.tf [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+      "image": "160372566358.dkr.ecr.ap-southeast-2.amazonaws.com/group3-wordpress:latest",
+      "cpu": 10,
+      "memory": 512,
+      "essential": true,
+
+      "secrets": [
+        { 
+          "name": "WORDPRESS_DB_HOST",
+          "valueFrom": "arn:aws:ssm:ap-southeast-2:160372566358:parameter/WORDPRESS_DB_HOST"
+        }
```
Comment (hsy3418) [inline wordpress-application/task-definition.tf]: Seems like it missing comma between these blocks

**Review by hsy3418 — CHANGES_REQUESTED** [review]

File: wordpress-application/task-definition.tf (hunk 2 above)
Comment (CarineOliveira) [reply wordpress-application/task-definition.tf:61]: kept the volumes out of the `containerDefinitions`

**Review by caiocezart — CHANGES_REQUESTED** [review]
good work 👏

**Review by caiocezart — APPROVED** [review]

File: wordpress-application/task-definition.tf [hunk 8]
```diff
@@ -0,0 +1,58 @@
+data "aws_iam_role" "ecs_role" {
```
Comment (caiocezart) [inline wordpress-application/task-definition.tf:5]: you need to create your own role.. this role only exists after running the ecs wizard on the console (on my case it failed as I have never used it before)

**Review by caiocezart — CHANGES_REQUESTED** [review]
need a ecs task execution role

File: wordpress-application/task-definition.tf (hunk 8 above)
Comment (CarineOliveira) [reply wordpress-application/task-definition.tf:5]: `aws_iam_policy_attachment` added

**Review by caiocezart — APPROVED** [review]

**Pull request closed without merging**